src/Socket_p.h

```diff
@@ -284,203 +243,192 @@ namespace Arcus
             {
-                message.size_recv += ret;
-                if(message.size_recv >= message.size)
+                current_message->received_size = current_message->received_size + result;
+                if(current_message->isComplete())
                 {
-                    if (message.valid) {
-                        message.state = MESSAGE_STATE_DISPATCH;
-                    }
-                    else
+                    if(!current_message->valid)
                     {
-                        delete[] message.data;
-                        message.data = nullptr;
-                        message.state = MESSAGE_STATE_INIT;
+                        current_message.reset();
+                        return;
                     }
+
+                    current_message->state = WireMessage::MessageState::Dispatch;
                 }
             }
         }
 
-        if (message.state == MESSAGE_STATE_DISPATCH)
-        {
-            handleMessage(message.type, message.size, message.data);
-            delete[] message.data;
-            message.data = nullptr;
-            message.state = MESSAGE_STATE_INIT;
-        }
-    }
-
-    int SocketPrivate::readInt32(int32_t *dest)
-    {
-        int32_t buffer;
-        int num = ::recv(socketId, reinterpret_cast<char*>(&buffer), 4, 0);
-
-        if(num != 4)
-        {
-            return -1;
-        }
-
-        *dest = ntohl(buffer);
-        return 0;
-    }
-
-    int SocketPrivate::readBytes(int size, char* dest)
-    {
-        int num = ::recv(socketId, dest, size, 0);
-        if(num == -1)
-        {
-            return -1;
-        }
-        else
+        if (current_message->state == WireMessage::MessageState::Dispatch)
         {
-            return num;
+            handleMessage(current_message);
+            current_message.reset();
         }
     }
 
-    void SocketPrivate::handleMessage(int type, int size, char* buffer)
+    // Parse and process a message received on the socket.
+    void Socket::Private::handleMessage(const std::shared_ptr<WireMessage>& wire_message)
     {
-        if(messageTypes.find(type) == messageTypes.end())
+        if(!message_types.hasType(wire_message->type))
         {
-            errorString = "Unknown message type";
+            error(ErrorCode::UnknownMessageTypeError, "Unknown message type");
             return;
         }
 
-        MessagePtr message = MessagePtr(messageTypes[type]->New());
-        google::protobuf::io::ArrayInputStream array(buffer, size);
+        MessagePtr message = message_types.createMessage(wire_message->type);
+
+        google::protobuf::io::ArrayInputStream array(wire_message->data, wire_message->size);
         google::protobuf::io::CodedInputStream stream(&array);
-        stream.SetTotalBytesLimit(500 * 1048576, 128 * 1048576); //Set size limit to 500MiB, warn at 128MiB
+        stream.SetTotalBytesLimit(500 * 1048576, 384 * 1048576); //Set size limit to 500MiB, warn at 368MiB
```

Comment:
Comment describes different value then what is actually set. Which one is correct (Do you want to warn at 368 or at 384?)